Speaker/Speaker.swift

```diff
@@ -0,0 +1,161 @@
+        get {
+            var volume: Float = 0
+            AudioQueueGetParameter(audioQueue, kAudioQueueParam_Volume, &volume)
+            return volume
+        }
+        set {
+            AudioQueueSetParameter(audioQueue, kAudioQueueParam_Volume, newValue)
+        }
+    }
+
+    internal weak var delegate: SpeakerDelegate?
+
+    internal init?(sampleRate: Float64) {
+
+        bufferSize = UInt32(sampleRate / 60)
+
+        var desc = AudioStreamBasicDescription(
+            mSampleRate:        sampleRate,
+            mFormatID:          kAudioFormatLinearPCM,
+            mFormatFlags:       kAudioFormatFlagIsFloat,
+            mBytesPerPacket:    8,
+            mFramesPerPacket:   1,
+            mBytesPerFrame:     8,
+            mChannelsPerFrame:  2,
+            mBitsPerChannel:    32,
+            mReserved:          0
+        )
+
+        let createError = AudioQueueNewOutput(&desc, Self.audioQueueOutputCallback, bridge(obj: self), nil, nil, 0, &audioQueue)
+
+        guard createError == noErr else {
+            assertionFailure("Error creating audio queue")
+            return nil
+        }
+
+        let isRunningListenerError = AudioQueueAddPropertyListener(audioQueue, kAudioQueueProperty_IsRunning, Self.startedListener, bridge(obj: self))
+
+        guard isRunningListenerError == noErr else {
+            assertionFailure("Error adding listener to audio queue")
+            return nil
+        }
+
+        for _ in 0..<Self.numberOfBuffers {
+            var audioQueueBuffer: AudioQueueBufferRef?
+            let error = AudioQueueAllocateBuffer(audioQueue, bufferSize, &audioQueueBuffer)
+
+            guard error == noErr else {
+                assertionFailure("Failed to create buffer")
+                return nil
+            }
+
+            guard let buffer = audioQueueBuffer else {
+                assertionFailure("Buffer is nil")
+                return nil
+            }
+
+            buffers.append(buffer)
+        }
+
+        let startError = AudioQueueStart(audioQueue, nil)
+
+        guard startError == noErr else {
+            assertionFailure("Failed to start")
+            return nil
+        }
+    }
+
+    deinit {
+        let error = AudioQueueDispose(audioQueue, true)
+        assert(error == noErr, "Error disposing audio queue")
+    }
+
+    private var audioQueue: AudioQueueRef!
+    private var buffers: [AudioQueueBufferRef] = []
+
+    private static let numberOfBuffers = 300
+    private let bufferSize: UInt32
+
+    private static let startedListener: AudioQueuePropertyListenerProc = { userData, audioQueue, propertyID in
+
+        guard propertyID == kAudioQueueProperty_IsRunning else {
+            assertionFailure("Wrong property")
+            return
+        }
+
+        var isRunning: UInt32 = 0
+        var size = UInt32(MemoryLayout.size(ofValue: UInt32.self))
+
+        let error = AudioQueueGetProperty(audioQueue, kAudioQueueProperty_IsRunning, &isRunning, &size)
```

Comment:
Maybe this should be wrapped in a function / getter?

Reply:
This is called from a separate thread, thats why I wanted to keep it local.